Simulator does not correctly handle non-debuggable types

```
error[E0277]: `Test` doesn't implement `Debug`
  --> dylib-examples/examples/22B40BD4.rs:80:36
   |
80 |                 format_item_debug: hydro_lang::__maybe_debug__!(),
   |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Debug` is not implemented for `Test`
   |
note: required by a bound in `IsDebug::<T>::format_debug`
  --> dylib-examples/examples/22B40BD4.rs:80:36
   |
80 |                 format_item_debug: hydro_lang::__maybe_debug__!(),
   |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                                    |
   |                                    required by a bound in this associated function
   |                                    required by this bound in `IsDebug::<T>::format_debug`
   = note: this error originates in the macro `hydro_lang::__maybe_debug__` (in Nightly builds, run with -Z macro-backtrace for more info)
```

--- a/hydro_lang/src/compile/trybuild/rewriters.rs
+++ b/hydro_lang/src/compile/trybuild/rewriters.rs
@@ -7,10 +7,13 @@ pub struct UseTestModeStaged<'a> {
 
 impl VisitMut for UseTestModeStaged<'_> {
     fn visit_type_path_mut(&mut self, i: &mut syn::TypePath) {
-        if let Some(first) = i.path.segments.first_mut()
+        if let Some(first) = i.path.segments.first()
             && first.ident == self.crate_name
+            && let Some(second) = i.path.segments.get(1)
+            && second.ident == "__staged"
         {
-            first.ident = syn::Ident::new("crate", Span::call_site());
+            i.path.segments.first_mut().unwrap().ident =
+                syn::Ident::new("crate", Span::call_site());
         }
 
         syn::visit_mut::visit_type_path_mut(self, i);
@@ -20,5 +23,20 @@ impl VisitMut for UseTestModeStaged<'_> {
         if i.ident == self.crate_name {
             i.ident = syn::Ident::new("crate", i.ident.span());
         }
+    }
+
+    fn visit_macro_mut(&mut self, i: &mut syn::Macro) {
+        // Rewrite types inside __maybe_debug__!(Type) invocations.
+        if i.path
+            .segments
+            .last()
+            .is_some_and(|s| s.ident == "__maybe_debug__")
+            && let Ok(mut ty) = i.parse_body::<syn::Type>()
+        {
+            self.visit_type_mut(&mut ty);
+            i.tokens = quote::quote! { #ty };
+        }
+
+        syn::visit_mut::visit_macro_mut(self, i);
     }
 }

--- a/hydro_lang/src/sim/builder.rs
+++ b/hydro_lang/src/sim/builder.rs
@@ -150,6 +150,7 @@ impl DfirBuilder for SimBuilder {
                 CollectionKind::Stream {
                     order,
                     retry: StreamRetry::ExactlyOnce,
+                    element_type,
                     ..
                 } => {
                     debug_assert!(in_location.is_top_level());
@@ -188,7 +189,7 @@ impl DfirBuilder for SimBuilder {
                                 to_release: None,
                                 output: #hoff_send_ident,
                                 batch_location: (#batch_location, #line, #caret),
-                                format_item_debug: #root::__maybe_debug__!(),
+                                format_item_debug: #root::__maybe_debug__!(#element_type),
                                 _order: std::marker::PhantomData,
                             })
                         ),
@@ -213,6 +214,8 @@ impl DfirBuilder for SimBuilder {
                 CollectionKind::KeyedStream {
                     value_order,
                     value_retry: StreamRetry::ExactlyOnce,
+                    key_type,
+                    value_type,
                     ..
                 } => {
                     debug_assert!(in_location.is_top_level());
@@ -251,7 +254,7 @@ impl DfirBuilder for SimBuilder {
                                 to_release: None,
                                 output: #hoff_send_ident,
                                 batch_location: (#batch_location, #line, #caret),
-                                format_item_debug: #root::__maybe_debug__!(),
+                                format_item_debug: #root::__maybe_debug__!((#key_type, #value_type)),
                                 _order: std::marker::PhantomData,
                             })
                         ),
@@ -273,7 +276,7 @@ impl DfirBuilder for SimBuilder {
                         None,
                     );
                 }
-                CollectionKind::Singleton { .. } => {
+                CollectionKind::Singleton { element_type, .. } => {
                     debug_assert!(in_location.is_top_level());
 
                     let hoff_id = self.next_hoff_id;
@@ -300,7 +303,7 @@ impl DfirBuilder for SimBuilder {
                                 #buffered_ident.clone(),
                                 #hoff_send_ident,
                                 (#batch_location, #line, #caret),
-                                #root::__maybe_debug__!(),
+                                #root::__maybe_debug__!(#element_type),
                             ))
                         ),
                     );
@@ -321,7 +324,11 @@ impl DfirBuilder for SimBuilder {
                         None,
                     );
                 }
-                CollectionKind::KeyedSingleton { .. } => {
+                CollectionKind::KeyedSingleton {
+                    key_type,
+                    value_type,
+                    ..
+                } => {
                     debug_assert!(in_location.is_top_level());
 
                     let hoff_id = self.next_hoff_id;
@@ -348,8 +355,8 @@ impl DfirBuilder for SimBuilder {
                                 #buffered_ident.clone(),
                                 #hoff_send_ident,
                                 (#batch_location, #line, #caret),
-                                #root::__maybe_debug__!(),
-                                #root::__maybe_debug__!(),
+                                #root::__maybe_debug__!(#key_type),
+                                #root::__maybe_debug__!(#value_type),
                             ))
                         ),
                     );
@@ -533,6 +540,7 @@ impl DfirBuilder for SimBuilder {
                     CollectionKind::Stream {
                         order: StreamOrder::NoOrder,
                         retry: StreamRetry::ExactlyOnce,
+                        element_type,
                         ..
                     },
                     CollectionKind::Stream {
@@ -570,7 +578,7 @@ impl DfirBuilder for SimBuilder {
                                 #buffered_ident.clone(),
                                 #hoff_send_ident,
                                 (#assume_location, #line, #caret),
-                                #root::__maybe_debug__!(),
+                                #root::__maybe_debug__!(#element_type),
                             ))
                         ),
                     );
@@ -605,6 +613,8 @@ impl DfirBuilder for SimBuilder {
                     CollectionKind::KeyedStream {
                         value_order: StreamOrder::NoOrder,
                         value_retry: StreamRetry::ExactlyOnce,
+                        key_type,
+                        value_type,
                         ..
                     },
                     CollectionKind::KeyedStream {
@@ -642,8 +652,8 @@ impl DfirBuilder for SimBuilder {
                                 #buffered_ident.clone(),
                                 #hoff_send_ident,
                                 (#assume_location, #line, #caret),
-                                #root::__maybe_debug__!(),
-                                #root::__maybe_debug__!(),
+                                #root::__maybe_debug__!(#key_type),
+                                #root::__maybe_debug__!(#value_type),
                             ))
                         ),
                     );
@@ -691,6 +701,7 @@ impl DfirBuilder for SimBuilder {
                     CollectionKind::Stream {
                         order: StreamOrder::NoOrder,
                         retry: StreamRetry::ExactlyOnce,
+                        element_type,
                         ..
                     },
                     CollectionKind::Stream {
@@ -724,7 +735,7 @@ impl DfirBuilder for SimBuilder {
                                 to_release: None,
                                 output: #hoff_send_ident,
                                 location: (#assume_location, #line, #caret),
-                                format_item_debug: #root::__maybe_debug__!(),
+                                format_item_debug: #root::__maybe_debug__!(#element_type),
                             })
                         ),
                     );
@@ -749,6 +760,8 @@ impl DfirBuilder for SimBuilder {
                     CollectionKind::KeyedStream {
                         value_order: StreamOrder::NoOrder,
                         value_retry: StreamRetry::ExactlyOnce,
+                        key_type,
+                        value_type,
                         ..
                     },
                     CollectionKind::KeyedStream {
@@ -782,7 +795,7 @@ impl DfirBuilder for SimBuilder {
                                 to_release: None,
                                 output: #hoff_send_ident,
                                 location: (#assume_location, #line, #caret),
-                                format_item_debug: #root::__maybe_debug__!(),
+                                format_item_debug: #root::__maybe_debug__!((#key_type, #value_type)),
                             })
                         ),
                     );

--- a/hydro_lang/src/sim/graph.rs
+++ b/hydro_lang/src/sim/graph.rs
@@ -619,8 +619,8 @@ fn compile_sim_graph_trybuild(
     cluster_max_sizes: SparseSecondaryMap<LocationKey, usize>,
     process_tick_graphs: BTreeMap<LocationId, DfirGraph>,
     cluster_tick_graphs: BTreeMap<LocationId, DfirGraph>,
-    extra_stmts_global: Vec<syn::Stmt>,
-    extra_stmts_cluster: BTreeMap<LocationId, Vec<syn::Stmt>>,
+    mut extra_stmts_global: Vec<syn::Stmt>,
+    mut extra_stmts_cluster: BTreeMap<LocationId, Vec<syn::Stmt>>,
     crate_name: &str,
     is_test: bool,
 ) -> syn::File {
@@ -639,6 +639,18 @@ fn compile_sim_graph_trybuild(
 
         dfir_expr
     };
+
+    if is_test {
+        extra_stmts_global.iter_mut().for_each(|stmt| {
+            UseTestModeStaged { crate_name }.visit_stmt_mut(stmt);
+        });
+
+        extra_stmts_cluster.values_mut().for_each(|stmts| {
+            stmts.iter_mut().for_each(|stmt| {
+                UseTestModeStaged { crate_name }.visit_stmt_mut(stmt);
+            })
+        });
+    }
 
     let process_dfir_exprs = process_graphs
         .into_iter()

--- a/hydro_lang/src/sim/runtime.rs
+++ b/hydro_lang/src/sim/runtime.rs
@@ -15,24 +15,46 @@ use crate::live_collections::stream::{NoOrder, Ordering, TotalOrder};
 pub type Hooks<Key> = HashMap<(Key, Option<u32>), Vec<Box<dyn SimHook>>>;
 pub type InlineHooks<Key> = HashMap<(Key, Option<u32>), Vec<Box<dyn SimInlineHook>>>;
 
+#[doc(hidden)]
 #[macro_export]
 #[doc(hidden)]
 macro_rules! __maybe_debug__ {
-    () => {{
-        trait NotDebug {
-            fn format_debug(&self) -> Option<String> {
-                None
-            }
+    ($type:ty) => {{
+        // Inherent-shadows-trait trick (same as `impls` crate), but for
+        // a function pointer. The const is stored as a raw `*const ()` to
+        // avoid mentioning the type in the const type (which would force
+        // bound resolution and break the trick). We transmute back at the end.
+
+        #[expect(clippy::allow_attributes, reason = "macro codegen")]
+        #[allow(dead_code, reason = "shadowing trick")]
+        fn no_debug<T>(_: &T) -> Option<String> {
+            None
+        }
+        fn yes_debug<T: std::fmt::Debug>(v: &T) -> Option<String> {
+            Some(format!("{:?}", v))
         }
 
-        impl<T> NotDebug for T {}
-        struct IsDebug<T>(std::marker::PhantomData<T>);
-        impl<T: std::fmt::Debug> IsDebug<T> {
-            fn format_debug(v: &T) -> Option<String> {
-                Some(format!("{:?}", v))
-            }
+        trait __Fallback {
+            const MAYBE_DEBUG_FN: *const () = no_debug::<()> as *const ();
         }
-        IsDebug::format_debug
+        impl<T: ?Sized> __Fallback for T {}
+
+        struct __Wrap<T>(std::marker::PhantomData<T>);
+
+        #[expect(clippy::allow_attributes, reason = "macro codegen")]
+        #[allow(dead_code, reason = "shadowing trick")]
+        impl<T: std::fmt::Debug> __Wrap<T> {
+            const MAYBE_DEBUG_FN: *const () = yes_debug::<T> as *const ();
+        }
+
+        // SAFETY: The pointer is either `no_debug::<()>` or `yes_debug::<$type>`.
+        // `no_debug` ignores its argument entirely, so the ABI is compatible
+        // regardless of the concrete type.
+        unsafe {
+            std::mem::transmute::<*const (), fn(&$type) -> Option<String>>(
+                <__Wrap<$type>>::MAYBE_DEBUG_FN,
+            )
+        }
     }};
 }
 
@@ -1119,5 +1141,31 @@ impl<K: Hash + Eq + Clone, V> SimHook for TopLevelKeyedStreamOrderHook<K, V> {
         } else {
             panic!("No decision to release");
         }
+    }
+}
+
+#[cfg(test)]
+mod maybe_debug_tests {
+    struct NotDebuggable;
+
+    #[derive(Debug)]
+    struct Debuggable;
+
+    #[test]
+    fn test_non_debug_type_returns_none() {
+        let fmt_fn: fn(&NotDebuggable) -> Option<String> = crate::__maybe_debug__!(NotDebuggable);
+        assert_eq!(fmt_fn(&NotDebuggable), None);
+    }
+
+    #[test]
+    fn test_debug_type_returns_some() {
+        let fmt_fn: fn(&Debuggable) -> Option<String> = crate::__maybe_debug__!(Debuggable);
+        assert_eq!(fmt_fn(&Debuggable), Some("Debuggable".to_owned()));
+    }
+
+    #[test]
+    fn test_primitive_debug() {
+        let fmt_fn: fn(&i32) -> Option<String> = crate::__maybe_debug__!(i32);
+        assert_eq!(fmt_fn(&42), Some("42".to_owned()));
     }
 }

--- a/hydro_lang/src/sim/tests/mod.rs
+++ b/hydro_lang/src/sim/tests/mod.rs
@@ -1,3 +1,4 @@
+use serde::{Deserialize, Serialize};
 use stageleft::q;
 
 use crate::live_collections::sliced::sliced;
@@ -256,4 +257,27 @@ fn trace_for_fuzzed_batching_sliced() {
 
     let log_str = String::from_utf8(log_out).unwrap();
     hydro_build_utils::assert_snapshot!(log_str);
+}
+
+#[derive(Serialize, Deserialize)]
+struct Test {}
+
+#[test]
+fn sim_batch_nondebuggable_type() {
+    let mut flow = FlowBuilder::new();
+    let node = flow.process::<()>();
+
+    let (in_send, input) = node.sim_input::<_, TotalOrder, _>();
+
+    let tick = node.tick();
+    let _out_recv = input
+        .batch(&tick, nondet!(/** test */))
+        .count()
+        .all_ticks()
+        .sim_output();
+
+    flow.sim().exhaustive(async || {
+        in_send.send(Test {});
+        let _: Vec<_> = _out_recv.collect().await;
+    });
 }

